File: ws01_plumbing/src/py04_param/py04_param/demo01_param_server.py


import rclpy
from rclpy.node import Node


class MyParamServer(Node):
    def __init__(self):
        # 如果要允许删除参数，需要提前声明
        super().__init__("param_server_py", allow_undeclared_parameters=True)
        self.get_logger().info("参数服务端 Python")

    def delcare_param(self):
        self.get_logger().info("---------- 增 ----------")
        self.declare_parameter("car_name", "tiger")
        self.declare_parameter("height", 3.45)
        self.declare_parameter("wheels", 4)

        pass

    def get_param(self):
        self.get_logger().info("---------- 查 ----------")
        car = self.get_parameter("car_name")
        self.get_logger().info("(%s = %s)" % (car.name, car.value))

        params = self.get_parameters(["car_name", "wheels"])
        for p in params:
            self.get_logger().info("(%s = %s)" % (p.name, p.value))
        pass

    def update_param(self):
        self.get_logger().info("---------- 改 ----------")
        self.set_parameters([rclpy.Parameter("car_name", value="hello")])
        pass

    def del_param(self):
        self.get_logger().info("---------- 删 ----------")
        self.undeclare_parameter("car_name")
        pass




def main():
    rclpy.init()
    node = MyParamServer()
    node.delcare_param()
    node.get_param()
    node.update_param()
    node.get_param()
    node.del_param()
    node.get_param()

    rclpy.spin(node)
    rclpy.shutdown()



if __name__ == '__main__':
    main()

File: ws01_plumbing/src/cpp04_param/src/demo01_param_server.cpp


#include "rclcpp/rclcpp.hpp"

class MyParamServer: public rclcpp::Node{
public:
    // 如果允许参数删除，需要通过NodeOption来声明
    MyParamServer():Node("my_param_server_node", rclcpp::NodeOptions().allow_undeclared_parameters(true)){
        RCLCPP_INFO(this->get_logger(), "参数服务端创建 CPP");
        
        // 增
        
        // 查

        // 改

        // 删

    }

    void declare_param(){
        RCLCPP_INFO(this->get_logger(), "------------- 增 -------------");

        this->declare_parameter("car_name", "tiger");
        this->declare_parameter("width", 1.66);
        this->declare_parameter("wheels", 5);

        this->set_parameter(rclcpp::Parameter("height", 1.88));
    }

    void get_param(){
        RCLCPP_INFO(this->get_logger(), "------------- 查 -------------");
        
        auto car = this->get_parameter("car_name");
        RCLCPP_INFO(this->get_logger(), "key = %s, val = %s", car.get_name().c_str(), car.as_string().c_str()); 

        auto params = this->get_parameters({"car_name", "wheels", "width"});
        for (auto &&param : params)
        {
            RCLCPP_INFO(this->get_logger(), "(%s = %s)", param.get_name().c_str(), param.value_to_string().c_str()); 
        }
        
        RCLCPP_INFO(this->get_logger(), "是否包含car_name   : %d", this->has_parameter("car_name")); 
        RCLCPP_INFO(this->get_logger(), "是否包含height     : %d", this->has_parameter("height")); 

    }

    void update_param(){
        RCLCPP_INFO(this->get_logger(), "------------- 改 -------------");

        this->set_parameter(rclcpp::Parameter("width", 1.75));
    }

    void del_param(){
        RCLCPP_INFO(this->get_logger(), "------------- 删 -------------");
        this->undeclare_parameter("car_name");
        RCLCPP_INFO(this->get_logger(), "是否包含car_name   : %d", this->has_parameter("car_name")); 
    }
};
int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<MyParamServer>();
    node->declare_param();
    node->get_param();
    // node->update_param();
    // node->get_param();
    // node->del_param();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}